### PR TITLE
Bug 1986392: Persist kubelet node name for OpenStack nodes

### PR DESCRIPTION
**Pull request by mdbooth**
Fixes: #2693

**- What I did**
This comprises 2 commits. The first commit simply allows --hostname-override to be set via the KUBELET_NODE_NAME environment variable. This environment variable is not set by default so will pass an empty string. Kubelet explicitly handles this as unset: [GetHostname()](https://github.com/kubernetes/kubernetes/blob/d9d4f0c69cbd3922dce7db924051f96e50524c38/pkg/util/node/node.go#L52-L70).

The second commit sets KUBELET_NODE_NAME on an OpenStack node to the unqualified instance name, which is compatible with the in-tree provider. It persists this value locally, and will not overwrite it on future invocations. This has 2 advantages:

Firstly, it fixes the upgrade bug.

Secondly, it allows for this upgrade-specific workaround to be removed in the future **if we also set providerID in kubelet.conf**, which we should probably also start doing. Because the value persists for nodes which were previously initialised with Node.name == _unqualified_, if we remove openstack-kubelet-nodename existing nodes will retain their existing node name, but new nodes will be initialised with FQDN hostname, consistent with other cloud providers.

**- How to verify it**
I have verified it with manual testing. If I can work out how to integrate an unmerged PR into my OpenStack test against the CCM repo I will also test it there.

**- Description for the changelog**
Persist kubelet node name for OpenStack nodes

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Changes seem reasonable to me, but we should definitely rope in the Node team to review this, CC @rphillips

**Comment by rphillips** [comment]
Had a chat with Joel and Matthew. This PR is going to be updated with AWS and Openstack support.

**Comment by rphillips** [comment]
Let's hold it until it's updated.

/hold

**Comment by mdbooth** [comment]
@rphillips has already added a hold. No idea if you can add 2 holds, but I'm going to add another while we continue to discuss the AWS changes.

/hold

**Comment by JoelSpeed** [comment]
Given that in the tests on AWS, I can see the nodes correctly joining the cluster with the correct names, I think this is working as expected for AWS

File: templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml
Comment (elmiko) [inline templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml:17]: would be nice to have a comment referring to the aws docs that reference this ip address, just so it's less confusing to a future person looking at this.

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml
Comment (elmiko) [inline templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:17]: similar comment here, although s/aws/openstack ;)

**Review by elmiko — COMMENTED** [review]
@mdbooth this generally makes sense to me, i just have a suggestion about adding a few more comments.

**Comment by mdbooth** [comment]
From the test jobs it looks like this is working correctly. Specifically I see that e2e-aws, e2e-aws-serial, and e2e-aws-upgrade-single-node all pass.

I'd like to work out why e2e-aws-upgrade-single-node, e2e-aws-disruptive, and preferably e2e-aws-techpreview-featuregate fail before pulling the trigger, but at first glance the failures don't look related.

**Comment by mdbooth** [comment]
aws-techpreview-featuregate is blocked by https://github.com/openshift/origin/pull/26349

**Comment by mdbooth** [comment]
e2e-aws-upgrade failed because one of the master nodes wasn't ready. Bizarrely the must-gather-extra also had no journal for it, but it did have output in the aws-console. Hard to tell what's going on there without logs. Lets see if it does it again.

/test e2e-aws-upgrade

**Comment by mdbooth** [comment]
This time e2e-aws-disruptive failed because:

```
fail [github.com/openshift/origin/test/extended/util/disruption/frontends/frontends.go:137]: Jul 30 18:58:36.279: Frontends were unreachable during disruption for at least 100.199263ms of 1h26m25s (0%):

Jul 30 18:14:55.648 E ns/openshift-authentication route/oauth-openshift Route stopped responding to GET requests over new connections
Jul 30 18:14:55.648 - 100ms E ns/openshift-authentication route/oauth-openshift Route is not responding to GET requests over new connections
Jul 30 18:14:55.749 I ns/openshift-authentication route/oauth-openshift Route started responding to GET requests over new connections

failed: (1h27m26s) 2021-07-30T18:59:35 "[sig-arch][Feature:ClusterUpgrade] Cluster should remain functional during upgrade [Disruptive] [Serial]"
```

However, by the time must-gather ran there were no degraded clusteroperators again. I'm going to call that an unrelated flake.

**Comment by mdbooth** [comment]
It looks like e2e-aws-disruptive failed because kube-apiserver-operator believed that the static pod kube-apiserver-ip-10-0-218-189 was not running:

```
    - lastTransitionTime: "2021-07-30T17:18:05Z"
      message: 'StaticPodsDegraded: pods "kube-apiserver-ip-10-0-218-189.us-west-1.compute.internal"
        not found'
      reason: StaticPods_Error
      status: "True"
      type: Degraded
```

It did actually start running on that host, though, only about 3 minutes later:

```
Jul 30 17:21:31.632315 ip-10-0-218-189 crio[1322]: time="2021-07-30 17:21:31.632265862Z" level=info msg="Started container b7090dd0b038c1173c36fabf0edd80e99a38fdef7db0b568618d48296ca37bec: openshift-kube-apiserver/kube-apiserver-ip-10-0-218-189.us-west-1.compute.internal/setup" id=af1b8f24-733f-45ca-beb3-b7838f91bb30 name=/runtime.v1alpha2.RuntimeService/StartContainer
```

No idea what that's about, but it doesn't seem to be hostname related.

File: templates/master/01-master-kubelet/_base/units/kubelet.service.yaml
Comment (mandre) [inline templates/master/01-master-kubelet/_base/units/kubelet.service.yaml:41]: `${KUBELET_NODE_NAME}` is possibly undefined, is it fine?

From the `e2e-metal-ipi` job I can see kubelet was started with `FLAG: --hostname-override=""` so I guess it's OK.

**Review by mandre — COMMENTED** [review]
You're modifying on-prem templates, let's manually trigger jobs for other on-prem platforms.

/test e2e-openstack
/test e2e-ovirt
/test e2e-vsphere

File: templates/master/01-master-kubelet/_base/units/kubelet.service.yaml
Comment (mdbooth) [reply templates/master/01-master-kubelet/_base/units/kubelet.service.yaml:41]: Yes, this is fine. Also passing an empty override works the same as passing no override, and is documented to do this.

File: templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml
Comment (mdbooth) [reply templates/common/aws/files/usr-local-bin-aws-kubelet-nodename.yaml:19]: Done

File: templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml
Comment (mdbooth) [reply templates/common/openstack/files/usr-local-bin-openstack-kubelet-nodename.yaml:19]: Done.

**Comment by JoelSpeed** [comment]
/lgtm

I appreciate that the tests don't explicitly pass on this PR, though we can see from the logs that the intension of this PR is working and that nodes are being named correctly, I think this is good to go

**Comment by mandre** [comment]
/hold
I would like to see e2e-openstack run on this PR first.
/test e2e-openstack

**Review by elmiko — COMMENTED** [review]
thanks for the updates @mdbooth 
lgtm from me

**Review by mandre — COMMENTED** [review]
/hold cancel
/lgtm

**Comment by mandre** [comment]
This should be good to go I believe, just need the final approval.
/assign @sinnykumari

**Comment by sinnykumari** [comment]
Before approving, would like to get final revision reviewed by node team.
@rphillips PTAL

**Comment by rphillips** [comment]
@sinnykumari lgtm

**Review by sinnykumari — APPROVED** [review]
Adding approval based on node team and others review.

**Pull request merged**
